File: client/automation.py
"""What the automation is doing, as one page you can read.

The office already ships every part of this answer, and that is exactly the
problem it has: the schedule is on the Pipeline card, the public path is on the
Webhooks card, and what the runner actually DID is 24 receipts buried on each of
72 desks. Nobody assembles them, so the question a person actually asks -- "is
the thing running, what did it just touch, and where is what it said" -- takes
three screens and a terminal.

This module is that assembly and nothing else. It reads no files, runs no
subprocesses and calls no network: `build()` is a pure function of the snapshot
being built around it. Everything it knows, some other part of the office
already measured, which is the point. A second measurement is a second number to
keep in step and the first one to go wrong.

THE LINK IS THE FEATURE
-----------------------
A row that says "commented on warehouse#284 an hour ago" and makes you go and find
it is a row that costs more than it saves. So every row carries two links:

  issue_url    the thread. Always present when the receipt names an issue,
               because it is built from `owner/name` and a number.
  comment_url  the exact comment the runner left, when the office knows it.

`comment_url` comes from the desk, not from the receipt: office-sync already
pulls each open issue's last comment, so when the runner still has the last word
the URL of the thing it said is right there. It is deliberately absent, rather
than approximated, in the two cases where it is not known:

  * a human replied after the runner, so the last comment is theirs. The runner's
    comment is still in that thread and the issue link is the honest way to it.
  * the issue is closed, or on a desk that is put away, so it is not in the
    snapshot at all.

An approximated deep link is worse than no deep link: it lands on the wrong
comment and reads exactly like the right one.

WHAT `activity` IS AND IS NOT
-----------------------------
It is the runner's receipts, filtered to the ones that name an issue. That is a
choice: a sweep writes one `survey` receipt per repo per pass, so the raw file is
92% "9 open: 0 to work, 9 waiting on a human" and reading it is how nobody ever
notices the four rows that mattered. The surveys are still counted, in
`reached`, so what was dropped is visible as a number rather than silently.

It is NOT a log of what the pipeline is doing right now. That is `now`, and it
comes from the pipeline source's reading of the live pid and log.
"""

from __future__ import annotations

from sources import _card

# One screen of history. The receipts file goes back weeks and a page nobody
# scrolls to the bottom of is a page with a bottom nobody has read.
MAX_ACTIVITY = 60

# The outcomes a receipt can carry, and what each one means for a person. This
# table IS the "how does it work" answer for the activity list: without it every
# row is a word from a bash script that only dispatch.sh's author can read.
OUTCOMES = {
    "landed": ("ok", "opened a PR that closes the issue"),
    "parked": ("dim", "left alone on purpose: this repo is configured to park"),
    "refused": ("warn", "would not touch it, and said why"),
    "deferred": ("dim", "ran out of run, will pick it up next sweep"),
    "timeout": ("bad", "hit its per-issue cap and was stopped"),
    "report-only": ("dim", "reviewed and commented; this repo grants no code change"),
    "dry-run": ("dim", "decided everything, changed nothing"),
    "no-access": ("warn", "no account here can push, so nothing was attempted"),
    "no-issues": ("dim", "nothing open to work on"),
    "caught-up": ("dim", "every open issue is waiting on a human"),
    "survey": ("dim", "counted the open issues and moved on"),
}

# The receipts that describe the SWEEP rather than an issue. Same set office-sync
# calls NOISE, for the same reason, and named separately here because this module
# drops them from a list rather than skipping them in a headline.
PER_REPO = {"survey", "no-issues", "caught-up", "no-access"}

# How the whole thing works, in the order the events happen. Written here rather
# than in a Swift string and a JavaScript string, because it is one explanation
# and two copies of an explanation drift the first time the mechanism does.
HOW = [
    "A launchd job wakes the runner on the interval below. It sweeps every repo "
    "the office has a desk for, one at a time.",
    "On each repo it lists the open issues and picks the ones the bot did NOT "
    "comment on last. An issue the bot spoke on last is waiting on you, and it "
    "is left alone until you reply. No label gates anything.",
    "It runs one agent lane against one issue at a time, capped in minutes. "
    "What it is allowed to do is that repo's capability: most can open a branch "
    "and a PR, none can merge, none can push to a default branch.",
    "Every issue it touches gets a comment carrying the bot marker, even when the "
    "lane failed. An issue with no marker would be retried forever.",
    "A webhook does the same thing without waiting for the hour: GitHub posts to "
    "the door the moment an issue or comment changes, and the door asks the "
    "runner to look at that one repo.",
]


def _tone(outcome: str) -> str:
    return OUTCOMES.get(outcome, ("", ""))[0]


def _means(outcome: str) -> str:
    return OUTCOMES.get(outcome, ("", ""))[1]


def _issue_index(stations: list) -> dict:
    """{(repo, "284"): issue row} for every open issue on every visible desk.

    Keyed on the issue number AS A STRING, because that is what a receipt holds:
    dispatch.sh writes whatever `gh` gave it, and comparing "284" to 284 is the
    kind of quiet miss that would empty this whole column without an error.
    """
    out = {}
    for station in stations or []:
        repo = str((station or {}).get("repo") or "")
        if not repo:
            continue
        for issue in (station.get("issues") or []):
            num = str((issue or {}).get("number") or "").strip()
            if num:
                out[(repo, num)] = issue
    return out


def _row(receipt: dict, issues: dict, now_iso: str) -> dict:
    repo = str(receipt.get("repo") or "")
    issue = str(receipt.get("issue") or "").strip()
    outcome = str(receipt.get("outcome") or "")
    at = str(receipt.get("at") or "")
    known = issues.get((repo, issue)) or {}
    # Only when the bot still has the last word does that URL point at what the
    # runner said. A human reply moves it, and a link to a human's comment
    # labelled as the runner's is a lie with an anchor on it.
    comment = str(known.get("last_word_url") or "") if known.get("bot_last") else ""
    return {
        "at": at,
        "ago": _card.ago(at) or "",
        "repo": repo,
        "issue": issue,
        "outcome": outcome,
        "tone": _tone(outcome),
        "means": _means(outcome),
        "detail": str(receipt.get("detail") or "")[:200],
        "title": str(known.get("title") or "")[:120],
        "issue_url": f"https://github.com/{repo}/issues/{issue}" if repo and issue else "",
        "comment_url": comment,
        "comment_at": str(known.get("last_word_at") or "") if comment else "",
    }


def build(by_repo: dict, stations: list, sections: dict, counts: dict,
          now_iso: str = "") -> dict:
    """The whole automation page, from what the snapshot already measured.

    `by_repo`  the runner's receipts, grouped and newest-first, from office-sync
    `stations` the desks, for issue titles and comment links
    `sections` the room's fixtures, for the pipeline's live state and the door's
    `counts`   the rolling 24h outcome tally office-sync already derived

    Every field the app and the phone draw is present in every return, including
    when the pipeline could not be read at all, so no renderer has to decide
    whether a missing key means "no" or means "nobody looked".
    """
    pipeline = (sections or {}).get("pipeline") or {}
    webhook = (sections or {}).get("webhook") or {}
    issues = _issue_index(stations)

    rows = []
    for repo_receipts in (by_repo or {}).values():
        for receipt in repo_receipts or []:
            if not isinstance(receipt, dict):
                continue
            if not str(receipt.get("issue") or "").strip():
                continue
            if str(receipt.get("outcome") or "") in PER_REPO:
                continue
            rows.append(receipt)
    rows.sort(key=lambda r: str(r.get("at") or ""), reverse=True)
    dropped = max(0, len(rows) - MAX_ACTIVITY)
    activity = [_row(r, issues, now_iso) for r in rows[:MAX_ACTIVITY]]

    cover = pipeline.get("covered") or {}
    state = str(pipeline.get("state") or "unknown")

    return {
        "state": state,
        "headline": _headline(pipeline, webhook, activity),
        "how": list(HOW),

        # WHEN it looks.
        "schedule": {
            "every": str(pipeline.get("every") or ""),
            "next_in": pipeline.get("next_in"),
            "next_at": pipeline.get("next_at"),
            "overdue": bool(pipeline.get("overdue")),
            "late_by": pipeline.get("late_by"),
            "enabled": pipeline.get("enabled"),
            "kill_switch": bool(pipeline.get("kill_switch")),
            "power": str(pipeline.get("power") or "unknown"),
            # dispatch defers the whole run on battery, so an hourly deferral
            # and an idle hour look identical without this.
            "deferring": bool(pipeline.get("deferring")),
            "last_full_run": pipeline.get("heartbeat"),
            "last_full_run_age_s": pipeline.get("heartbeat_age_s"),
        },

        # WHETHER it is looking right now.
        "now": {
            "running": bool(pipeline.get("running")),
            "for": pipeline.get("running_for"),
            "doing": str(pipeline.get("doing") or ""),
            "last_said": str(pipeline.get("last_said") or "")[:400],
            "last_said_at": pipeline.get("last_said_at"),
            "last_said_age_s": pipeline.get("last_said_age_s"),
            "stale_pid": pipeline.get("stale_pid"),
            "detail": str(pipeline.get("detail") or ""),
        },

        # The other way in: GitHub telling us, instead of us asking.
        "trigger": {
            "state": str(webhook.get("state") or "unknown"),
            "reachable": bool(webhook.get("state") == "ok"),
            "deliveries": webhook.get("events_total"),
            "today": webhook.get("events_today"),
            "last_at": webhook.get("last_at"),
            "last_age_s": webhook.get("last_age_s"),
            "runs_today": webhook.get("runs_today"),
            "queued": len(webhook.get("queued") or []),
            "detail": str(webhook.get("detail") or ""),
            "blocked_by": str(webhook.get("blocked_by") or ""),
        },

        # WHAT it reached, and what this list is not showing.
        "reached": {
            "repos": cover.get("repos"),
            "receipts": cover.get("receipts"),
            "window": "24h",
            "state": str(cover.get("state") or "unknown"),
        },
        "counts": dict(counts or {}),
        "activity": activity,
        # Never a silent truncation: a list capped without saying so reads as
        # "that is everything that happened".
        "activity_dropped": dropped,
    }


def _headline(pipeline: dict, webhook: dict, activity: list) -> str:
    """The one sentence, and it says the WORST true thing first.

    Order matters and is not aesthetic. A run in flight is the most present
    fact; a scheduler that stopped firing is the most expensive one to not
    notice; and "nothing happened" is only reassuring once you know both.
    """
    state = str(pipeline.get("state") or "")
    if state not in ("ok", "off"):
        return _card.clip(pipeline.get("detail") or f"the pipeline is {state}", 79)
    # A run in flight outranks the kill switch, exactly as it does on the
    # Pipeline card. The switch stops the NEXT run; it does not reach into one
    # already going, and a page that says "switched off" while a lane is pushing
    # a branch is the more dangerous of the two sentences.
    if pipeline.get("running"):
        doing = str(pipeline.get("doing") or "").strip()
        return _card.clip(f"running now: {doing}" if doing else "a run is in flight", 79)
    if pipeline.get("kill_switch"):
        return "switched off at the kill switch; nothing will run until it is removed"
    if state == "off":
        return _card.clip(pipeline.get("detail") or "the job is off", 79)
    if pipeline.get("overdue"):
        return _card.clip(
            f"overdue by {pipeline.get('late_by') or 'a while'}; it should have looked already", 79)
    if pipeline.get("deferring"):
        return "on battery, so every run is deferring without doing anything"
    touched = len(activity)
    nxt = pipeline.get("next_in") or "soon"
    if not touched:
        return _card.clip(f"idle, next look {nxt}; no issue touched in the last day", 79)
    return _card.clip(
        f"idle, next look {nxt}; {touched} {_card.plural(touched, 'issue')} touched recently", 79)

File: app/Office/Model/Automation.swift
import Foundation

/// The automation, as one page: what the schedule is, whether it is running,
/// what it touched, and where what it said is.
///
/// Every value here was measured by the server when it built the snapshot. This
/// file decodes and nothing else: it does no arithmetic on ages, picks no
/// headline, and never decides whether a state is bad. All of that lives in
/// `client/automation.py`, so the phone page and this app say the same sentence
/// about the same machine rather than two sentences that agree until they do
/// not.
///
/// Decoding is lenient the whole way down, for the reason the rest of
/// `Models.swift` is: a server that has not learned a field yet must draw a page
/// with a gap in it, never no page at all.

public struct Automation: Decodable, Equatable {
    public var state: String = "unknown"
    public var headline: String = ""
    /// How the whole thing works, in the order the events happen. From the
    /// server, so there is one copy of the explanation rather than one here and
    /// one in the phone's JavaScript.
    public var how: [String] = []
    public var schedule: Schedule = Schedule()
    public var now: Current = Current()
    public var trigger: Trigger = Trigger()
    public var reached: Reached = Reached()
    public var activity: [Activity] = []
    /// How many rows the server left off the end. Drawn, always: a list capped
    /// in silence reads as "that is everything that happened".
    public var activityDropped: Int = 0

    public var isEmpty: Bool { headline.isEmpty && activity.isEmpty }

    /// Whether anything on this page wants a person. The kill switch does not:
    /// somebody switched it off, which is a decision and not a fault.
    public var needsSomebody: Bool {
        schedule.overdue || schedule.deferring || now.stalePid != nil
            || !trigger.blockedBy.isEmpty || state == "unreadable"
    }

    enum CodingKeys: String, CodingKey {
        case state, headline, how, schedule, now, trigger, reached, activity
        case activityDropped = "activity_dropped"
    }

    public init() {}

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        state = c.str(.state) ?? "unknown"
        headline = c.str(.headline) ?? ""
        how = c.list(.how, String.self)
        schedule = ((try? c.decodeIfPresent(Schedule.self, forKey: .schedule)) ?? nil) ?? Schedule()
        now = ((try? c.decodeIfPresent(Current.self, forKey: .now)) ?? nil) ?? Current()
        trigger = ((try? c.decodeIfPresent(Trigger.self, forKey: .trigger)) ?? nil) ?? Trigger()
        reached = ((try? c.decodeIfPresent(Reached.self, forKey: .reached)) ?? nil) ?? Reached()
        // One row at a time: a receipt that will not decode loses its own row,
        // never the whole history.
        activity = c.list(.activity, Lenient<Activity>.self).compactMap(\.value)
        activityDropped = c.int(.activityDropped) ?? 0
    }

    // MARK: - when it looks

    public struct Schedule: Decodable, Equatable {
        public var every: String = ""
        public var nextIn: String = ""
        public var overdue: Bool = false
        public var lateBy: String = ""
        public var enabled: Bool?
        public var killSwitch: Bool = false
        public var power: String = "unknown"
        /// dispatch exits before doing anything on battery, so an hourly
        /// deferral and an idle hour look identical without this.
        public var deferring: Bool = false
        public var lastFullRun: String = ""
        public var lastFullRunAge: Int?

        enum CodingKeys: String, CodingKey {
            case every, overdue, power, deferring
            case nextIn = "next_in", lateBy = "late_by", killSwitch = "kill_switch"
            case enabled
            case lastFullRun = "last_full_run", lastFullRunAge = "last_full_run_age_s"
        }

        public init() {}

        public init(from decoder: Decoder) throws {
            let c = try decoder.container(keyedBy: CodingKeys.self)
            every = c.str(.every) ?? ""
            nextIn = c.str(.nextIn) ?? ""
            overdue = c.bool(.overdue) ?? false
            lateBy = c.str(.lateBy) ?? ""
            enabled = c.bool(.enabled)
            killSwitch = c.bool(.killSwitch) ?? false
            power = c.str(.power) ?? "unknown"
            deferring = c.bool(.deferring) ?? false
            lastFullRun = c.str(.lastFullRun) ?? ""
            lastFullRunAge = c.int(.lastFullRunAge)
        }

        /// "every hour, next look in 43 minutes". One line, and it says overdue
        /// rather than "in 0 minutes" when the scheduler has stopped firing.
        public var line: String {
            var parts: [String] = []
            if !every.isEmpty { parts.append("every \(every)") }
            if overdue {
                parts.append("overdue by \(lateBy.isEmpty ? "a while" : lateBy)")
            } else if !nextIn.isEmpty {
                parts.append("next look \(nextIn)")
            }
            return parts.joined(separator: ", ")
        }
    }

    // MARK: - whether it is looking right now

    public struct Current: Decodable, Equatable {
        public var running: Bool = false
        public var forHowLong: String = ""
        public var doing: String = ""
        public var lastSaid: String = ""
        public var lastSaidAge: Int?
        public var detail: String = ""
        /// A pid file naming a process that is not a live dispatch.sh. Never
        /// reported as a run: that is the whole point of it being here.
        public var stalePid: String?

        enum CodingKeys: String, CodingKey {
            case running, doing, detail
            case forHowLong = "for"
            case lastSaid = "last_said", lastSaidAge = "last_said_age_s"
            case stalePid = "stale_pid"
        }

        public init() {}

        public init(from decoder: Decoder) throws {
            let c = try decoder.container(keyedBy: CodingKeys.self)
            running = c.bool(.running) ?? false
            forHowLong = c.str(.forHowLong) ?? ""
            doing = c.str(.doing) ?? ""
            lastSaid = c.str(.lastSaid) ?? ""
            lastSaidAge = c.int(.lastSaidAge)
            detail = c.str(.detail) ?? ""
            let stale = (try? c.decodeIfPresent(StalePid.self, forKey: .stalePid)) ?? nil
            stalePid = stale?.why.isEmpty == false ? stale?.why : nil
        }

        private struct StalePid: Decodable {
            var why: String = ""
            enum CodingKeys: String, CodingKey { case why }
            init(from decoder: Decoder) throws {
                why = (try decoder.container(keyedBy: CodingKeys.self)).str(.why) ?? ""
            }
        }
    }

    // MARK: - the other way in

    public struct Trigger: Decodable, Equatable {
        public var state: String = "unknown"
        public var reachable: Bool = false
        public var deliveries: Int?
        public var today: Int?
        public var lastAt: String = ""
        public var lastAge: Int?
        public var runsToday: Int?
        public var queued: Int = 0
        public var detail: String = ""
        /// The one actionable line when nothing is arriving. Empty when nothing
        /// is proven to be blocking, never absent.
        public var blockedBy: String = ""

        enum CodingKeys: String, CodingKey {
            case state, reachable, deliveries, today, queued, detail
            case lastAt = "last_at", lastAge = "last_age_s"
            case runsToday = "runs_today", blockedBy = "blocked_by"
        }

        public init() {}

        public init(from decoder: Decoder) throws {
            let c = try decoder.container(keyedBy: CodingKeys.self)
            state = c.str(.state) ?? "unknown"
            reachable = c.bool(.reachable) ?? false
            deliveries = c.int(.deliveries)
            today = c.int(.today)
            lastAt = c.str(.lastAt) ?? ""
            lastAge = c.int(.lastAge)
            runsToday = c.int(.runsToday)
            queued = c.int(.queued) ?? 0
            detail = c.str(.detail) ?? ""
            blockedBy = c.str(.blockedBy) ?? ""
        }
    }

    // MARK: - what it reached

    public struct Reached: Decodable, Equatable {
        public var repos: Int?
        public var receipts: Int?
        public var window: String = "24h"
        public var state: String = "unknown"

        enum CodingKeys: String, CodingKey { case repos, receipts, window, state }

        public init() {}

        public init(from decoder: Decoder) throws {
            let c = try decoder.container(keyedBy: CodingKeys.self)
            repos = c.int(.repos)
            receipts = c.int(.receipts)
            window = c.str(.window) ?? "24h"
            state = c.str(.state) ?? "unknown"
        }
    }

    // MARK: - one thing the runner did

    public struct Activity: Decodable, Equatable, Identifiable {
        public var at: String = ""
        public var ago: String = ""
        public var repo: String = ""
        public var issue: String = ""
        public var outcome: String = ""
        public var tone: String = ""
        /// What that outcome means, in words a person who has never read
        /// dispatch.sh can act on.
        public var means: String = ""
        public var detail: String = ""
        public var title: String = ""
        public var issueURL: String = ""
        /// The exact comment the runner left, when the office knows it. Empty
        /// when a human replied after it, which moves the last comment: a deep
        /// link to somebody else's words labelled as the runner's is worse than
        /// no deep link.
        public var commentURL: String = ""
        public var commentAt: String = ""

        /// Stable across polls: a receipt is one repo, one issue, one moment.
        public var id: String { "\(repo)#\(issue)@\(at)" }

        public var hasComment: Bool { !commentURL.isEmpty }
        public var link: URL? { URL(string: hasComment ? commentURL : issueURL) }

        enum CodingKeys: String, CodingKey {
            case at, ago, repo, issue, outcome, tone, means, detail, title
            case issueURL = "issue_url", commentURL = "comment_url", commentAt = "comment_at"
        }

        public init() {}

        public init(from decoder: Decoder) throws {
            let c = try decoder.container(keyedBy: CodingKeys.self)
            at = c.str(.at) ?? ""
            ago = c.str(.ago) ?? ""
            repo = c.str(.repo) ?? ""
            issue = c.str(.issue) ?? ""
            outcome = c.str(.outcome) ?? ""
            tone = c.str(.tone) ?? ""
            means = c.str(.means) ?? ""
            detail = c.str(.detail) ?? ""
            title = c.str(.title) ?? ""
            issueURL = c.str(.issueURL) ?? ""
            commentURL = c.str(.commentURL) ?? ""
            commentAt = c.str(.commentAt) ?? ""
        }
    }
}
